### DISCUSSION TITLE
Monkeypatch not working on "from module import X"?

### DISCUSSION THREAD

**Discussion by thomjah-computas** [pytest-dev/pytest#10216] [Q&A] [answered]
I'm trying to mock `uuid.uuid4()` to give a constant value in the test case, but it is not triggering when the code uses `from uuid import uuid4`.

```
from uuid import UUID, uuid4
import uuid

import pytest


def make_plain_uuid():
    return uuid4()


def make_modular_uuid():
    return uuid.uuid4()


def test_patch_uuid_plain(monkeypatch: pytest.MonkeyPatch):
    fixed_uuid = UUID("c559e06e-f0d7-4047-abf3-20faab54dfb6")

    def mock_uuid():
        return fixed_uuid

    monkeypatch.setattr(uuid, "uuid4", mock_uuid, raising=True)

    actual = make_plain_uuid()

    assert actual == fixed_uuid


def test_patch_uuid_modular(monkeypatch: pytest.MonkeyPatch):
    fixed_uuid = UUID("c559e06e-f0d7-4047-abf3-20faab54dfb6")

    def mock_uuid():
        return fixed_uuid

    monkeypatch.setattr(uuid, "uuid4", mock_uuid, raising=True)

    actual = make_modular_uuid()

    assert actual == fixed_uuid


def test_is_uuid4_the_same():
    from_module_id = id(uuid4)
    modular_id = id(uuid.uuid4)

    assert from_module_id == modular_id

```

`test_patch_uuid_plain` fails as it doesn't patch the function. Is this a known problem with the monkeypatch, and is there some way to fix it? 

It is running `platform darwin -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0`

**Comment by RonnyPfannschmidt** [comment, marked as answer]
That's expected behaviour, monkeypatch patches where you tell it, not everywhere it could possibly find something

**Reply by thomjah-computas** [reply]
Is there a way to patch uuid4() when it's imported as `from uuid import uuid4` in the application code?

**Reply by nicoddemus** [reply]
You will need to patch the application module, not `uuid`, say if the code that calls `uuid4` lives in `app.utils`, you need to call `monkeypatch.setattr(app.utils, "uuid4", ...)`.

There is a complete explanation in the [mock docs](https://docs.python.org/3/library/unittest.mock.html#where-to-patch), which apply for `monkeypatch.setattr` as well.

**Reply by thomjah-computas** [reply]
Thanks, that worked. Maybe it should be demonstrated in the documentation https://docs.pytest.org/en/7.1.x/how-to/monkeypatch.html . It's not so obvious how to fix it.

**Reply by nicoddemus** [reply]
Indeed, opened https://github.com/pytest-dev/pytest/pull/10217 for that. 👍
